### PR TITLE
feat: deployer service skeleton

### PR DESCRIPTION
**Pull request by bmoxb**

### CODE REVIEW

File: deployer/src/tower_service/mod.rs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    deployment_manager: DeploymentManager,
+    persistence: Persistence,
+}
+
+impl Deployer {
+    pub async fn new() -> Self {
+        let persistence = Persistence::new().await;
+
+        let cpus = num_cpus::get();
+        let pipeline_count = (cpus + 2) / 3; // TODO: How many is suitable?
```
Comment (brokad) [inline deployer/src/tower_service/mod.rs:32]: I don't believe this should be necessary. If we want all threads in `tokio`'s pool to be used, don't we just have to `::spawn` the `handle_queue` bit of pipeline steps right after a successful resolve of `recv.recv()`? IOW, when a step gets a new input, spawn the handler and move on to listening to the queue again.

File: deployer/src/deployment/mod.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        let (queue_send, queue_recv) = mpsc::channel(QUEUE_BUFFER_SIZE);
+        let (run_send, run_recv) = mpsc::channel(RUN_BUFFER_SIZE);
+
+        let run_send_clone = run_send.clone();
+        let persistence_clone = persistence.clone();
+
+        tokio::spawn(
+            async move { queue::task(ident, queue_recv, run_send_clone, persistence).await },
+        );
+        tokio::spawn(async move { run::task(ident, run_recv, kill_send, persistence_clone).await });
```
[the remark this answers is not in the document]
Comment (bmoxb) [reply deployer/src/deployment/mod.rs]: This is a cool idea - I'll look into it though maybe outside of the scope of this PR?

File: deployer/src/tower_service/mod.rs (hunk 1 above)
Comment (bmoxb) [reply deployer/src/tower_service/mod.rs]: Yeah I agree - have made it just a single pipeline with a 'fanning out' approach (i.e., continuously reading from channel and spawn as many tasks as necessary).

File: deployer/src/deployment/run.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+pub async fn task(
+    ident: usize,
+    mut recv: RunReceiver,
+    kill_send: KillSender,
+    persistence: Persistence,
+) {
+    log::info!("Run task {ident} started");
+
+    while let Some(mut built) = recv.recv().await {
```
[the remark this answers is not in the document]
Comment (bmoxb) [reply deployer/src/deployment/run.rs]: Sorry, maybe I'm missing something obvious but I'm not sure what exactly you're referring to? Can make the operation cancellable with `tokio::select` on some signal future quite easily - I'm not sure what in the `futures` crate would be needed?

[the remark this answers is not in the document]
Comment (brokad) [reply deployer/src/deployment/run.rs]: That works too, I was just giving a suggestion :smile:

**Pull request merged**